// @mariozechner/pi-ai/dist/utils/oauth/index.js
/**
 * OAuth credential management for AI providers.
 */
export { anthropicOAuthProvider, loginAnthropic, refreshAnthropicToken } from "./anthropic.js";
export { geminiCliOAuthProvider, loginGeminiCli, refreshGoogleCloudToken } from "./google-gemini-cli.js";
export { loginOpenAICodex, openaiCodexOAuthProvider, refreshOpenAICodexToken } from "./openai-codex.js";
export * from "./types.js";
import { anthropicOAuthProvider } from "./anthropic.js";
import { geminiCliOAuthProvider } from "./google-gemini-cli.js";
import { openaiCodexOAuthProvider } from "./openai-codex.js";
const BUILT_IN_OAUTH_PROVIDERS = [
    anthropicOAuthProvider,
    geminiCliOAuthProvider,
    openaiCodexOAuthProvider,
];
const oauthProviderRegistry = new Map(BUILT_IN_OAUTH_PROVIDERS.map((provider) => [provider.id, provider]));
export function getOAuthProvider(id) {
    return oauthProviderRegistry.get(id);
}
export function registerOAuthProvider(provider) {
    oauthProviderRegistry.set(provider.id, provider);
}
export function unregisterOAuthProvider(id) {
    const builtInProvider = BUILT_IN_OAUTH_PROVIDERS.find((provider) => provider.id === id);
    if (builtInProvider) {
        oauthProviderRegistry.set(id, builtInProvider);
        return;
    }
    oauthProviderRegistry.delete(id);
}
export function resetOAuthProviders() {
    oauthProviderRegistry.clear();
    for (const provider of BUILT_IN_OAUTH_PROVIDERS) {
        oauthProviderRegistry.set(provider.id, provider);
    }
}
export function getOAuthProviders() {
    return Array.from(oauthProviderRegistry.values());
}
export function getOAuthProviderInfoList() {
    return getOAuthProviders().map((p) => ({
        id: p.id,
        name: p.name,
        available: true,
    }));
}
export async function refreshOAuthToken(providerId, credentials) {
    const provider = getOAuthProvider(providerId);
    if (!provider) {
        throw new Error(`Unknown OAuth provider: ${providerId}`);
    }
    return provider.refreshToken(credentials);
}
export async function getOAuthApiKey(providerId, credentials) {
    const provider = getOAuthProvider(providerId);
    if (!provider) {
        throw new Error(`Unknown OAuth provider: ${providerId}`);
    }
    let creds = credentials[providerId];
    if (!creds) {
        return null;
    }
    if (Date.now() >= creds.expires) {
        try {
            creds = await provider.refreshToken(creds);
        }
        catch (_error) {
            throw new Error(`Failed to refresh OAuth token for ${providerId}`);
        }
    }
    const apiKey = provider.getApiKey(creds);
    return { newCredentials: creds, apiKey };
}
